File: frontend/src/api/projects.ts
import axios from 'axios'

const api = axios.create({ baseURL: '/api/v1', timeout: 30000 })

export interface NovelProject {
  id: string
  title: string
  description?: string
  genre?: string
  status: 'draft' | 'published' | 'archived'
  content?: string
  word_count: number
  created_at: string
  updated_at: string
  user_id?: string
}

export interface ProjectCreate {
  title: string
  description?: string
  genre?: string
  content?: string
  status?: string
}

export interface ProjectUpdate {
  title?: string
  description?: string
  genre?: string
  content?: string
  status?: string
}

export const getProjects = (status?: string) =>
  api.get<NovelProject[]>('/projects', { params: status ? { status } : {} }).then(r => r.data)

export const getProject = (id: string) =>
  api.get<NovelProject>(`/projects/${id}`).then(r => r.data)

export const createProject = (body: ProjectCreate) =>
  api.post<NovelProject>('/projects', body).then(r => r.data)

export const updateProject = (id: string, body: ProjectUpdate) =>
  api.put<NovelProject>(`/projects/${id}`, body).then(r => r.data)

export const deleteProject = (id: string) =>
  api.delete(`/projects/${id}`)

export const exportProject = async (id: string, format: 'markdown' | 'pdf' | 'epub', title: string) => {
  const res = await api.post(`/export/projects/${id}/${format}`, {}, { responseType: 'blob' })
  const ext = format === 'markdown' ? 'md' : format
  const url = URL.createObjectURL(res.data)
  const a = document.createElement('a')
  a.href = url
  a.download = `${title}.${ext}`
  a.click()
  URL.revokeObjectURL(url)
}

File: backend/api/routes/projects.py
"""
Project CRUD API routes
"""
import uuid
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import create_engine
from sqlalchemy.orm import Session, sessionmaker

from config.settings import settings
from database.models import Base, NovelProject

router = APIRouter(prefix="/projects", tags=["projects"])

engine = create_engine(settings.DATABASE_URL, connect_args={"check_same_thread": False} if "sqlite" in settings.DATABASE_URL else {})
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


class ProjectCreate(BaseModel):
    title: str
    description: Optional[str] = None
    genre: Optional[str] = None
    content: Optional[str] = None
    status: Optional[str] = "draft"


class ProjectUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    genre: Optional[str] = None
    content: Optional[str] = None
    status: Optional[str] = None


def project_to_dict(p: NovelProject) -> dict:
    return {
        "id": p.id,
        "title": p.title,
        "description": p.description,
        "genre": p.genre,
        "status": p.status,
        "content": p.content,
        "word_count": p.word_count,
        "created_at": p.created_at.isoformat() if p.created_at else None,
        "updated_at": p.updated_at.isoformat() if p.updated_at else None,
        "user_id": p.user_id,
    }


@router.post("", status_code=201)
def create_project(body: ProjectCreate, db: Session = Depends(get_db)):
    project = NovelProject(
        id=str(uuid.uuid4()),
        title=body.title,
        description=body.description,
        genre=body.genre,
        content=body.content,
        status=body.status or "draft",
        word_count=len(body.content.split()) if body.content else 0,
    )
    db.add(project)
    db.commit()
    db.refresh(project)
    return project_to_dict(project)


@router.get("")
def list_projects(status: Optional[str] = None, db: Session = Depends(get_db)):
    query = db.query(NovelProject)
    if status:
        query = query.filter(NovelProject.status == status)
    return [project_to_dict(p) for p in query.all()]


@router.get("/{project_id}")
def get_project(project_id: str, db: Session = Depends(get_db)):
    project = db.query(NovelProject).filter(NovelProject.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project_to_dict(project)


@router.put("/{project_id}")
def update_project(project_id: str, body: ProjectUpdate, db: Session = Depends(get_db)):
    project = db.query(NovelProject).filter(NovelProject.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(project, field, value)
    if body.content is not None:
        project.word_count = len(body.content.split())
    db.commit()
    db.refresh(project)
    return project_to_dict(project)


@router.delete("/{project_id}", status_code=204)
def delete_project(project_id: str, db: Session = Depends(get_db)):
    project = db.query(NovelProject).filter(NovelProject.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    db.delete(project)
    db.commit()
